nixos/modules/programs/fish.nix

```diff
@@ -79,31 +103,72 @@ in
     environment.etc."fish/foreign-env/loginShellInit".text = cfge.loginShellInit;
     environment.etc."fish/foreign-env/interactiveShellInit".text = cfge.interactiveShellInit;
 
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
+      # avoid clobbering the environment if it's been set by a parent shell
+
+      # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+      # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+      set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $__fish_datadir/functions
+      
+      # source the NixOS environment config
+      fenv source ${config.system.build.setEnvironment}
+
+      # clear fish_function_path so that it will be correctly set when we return to $__fish_datadir/config.fish
+      set -e fish_function_path
+    '';
+
     environment.etc."fish/config.fish".text = ''
       # /etc/fish/config.fish: DO NOT EDIT -- this file has been generated automatically.
 
-      set fish_function_path $fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions
-
-      fenv source ${config.system.build.setEnvironment} > /dev/null ^&1
-      fenv source /etc/fish/foreign-env/shellInit > /dev/null
+      # if our parent shell didn't source the general config, do it
+      if not set -q __fish_nixos_general_config_sourced
+        set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $fish_function_path
+        fenv source /etc/fish/foreign-env/shellInit > /dev/null
+        set -e fish_function_path[1]
+        
+        ${cfg.shellInit}
 
-      ${cfg.shellInit}
+        # and leave a note to our children to spare them the same work
+        set -gx __fish_nixos_general_config_sourced 1
+      end
 
-      if status --is-login
+      # if our parent shell didn't source the interactive config, do it
+      status --is-login; and not set -q __fish_nixos_interactive_config_sourced
+      and begin
+        set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $fish_function_path
         fenv source /etc/fish/foreign-env/loginShellInit > /dev/null
+        set -e fish_function_path[1]
+        
         ${cfg.loginShellInit}
+
+        # and leave a note to our children to spare them the same work
+        set -gx __fish_nixos_interactive_config_sourced 1
       end
 
-      if status --is-interactive
+      # if our parent shell didn't source the login config, do it
+      status --is-interactive; and not set -q __fish_nixos_login_config_sourced
```

Comment:
ay ay, this is mixed up in several ways. Thanks for pointing that out. I'll get that fixed right away.